### PR TITLE
Add optional lib_suffix

### PR DESCRIPTION
**Pull request by ObiWahn**
**What:**

This patch should allow for easier building of openssl windows libraries like they can be found on:
https://www.npcglib.org/~stathis/blog/precompiled-openssl/

**Why:**

We have currently the situation that we want to use a recent windows lib that is build with Visual Studio 2017. But there are only precompiled libraries that are not up to date or not compiled with Visual Studio 2017. First
I tried to apply the patches provided by _stathis_ (in the link above) to a more recent version but failed. So I came up with this patch that seems to solve our problem but most probably still needs a lot of reviewing and work to get it in the official repository.

I think it is important that people are able to build and use a recent openssl versions on windows that works together with tools like CMake. Of course I could have changed https://github.com/Kitware/CMake/blob/master/Modules/FindOpenSSL.cmake and introduced more places openssl is searched for. I did decide against it because the situation on windows concerning library locations is bad enough.

Please let me know what needs to be changed. Please note that I am a total perl newbie:)

### CODE REVIEW

File: Configurations/00-base-templates.conf [hunk 1]
```diff
@@ -43,6 +43,8 @@ my %targets=(
 	shared_ldflag	=> "",
 	shared_rcflag	=> "",
 	shared_extension	=> "",
+	
+	lib_suffix => "",
```
Comment (levitte) [inline Configurations/00-base-templates.conf]: You call it lib_suffix here but win_suffix in 10-main.conf.  I think you need to decide what it will be and use that consistently.

**Comment by dot-asm** [comment]
There is a hook introduced in #4760 that aims for similar kind of thing. I mean #4760 is quite Unix-centric, but it would probably be more appropriate to re-use it.

Secondly, and more importantly, I [for one] can't find the suggested modification to 10-main.cf appropriate. I mean adding hooks is fine, adding project-specific (or maintaining separately) .cf files is fine, but just dropping another file naming convention on everybody is hardly appropriate.

File: Configurations/windows-makefile.tmpl [hunk 2]
```diff
@@ -419,7 +421,7 @@ install_dev:
 	@"$(PERL)" "$(SRCDIR)\util\mkdir-p.pl" "$(libdir)"
 	@"$(PERL)" "$(SRCDIR)\util\copy.pl" $(INSTALL_LIBS) "$(libdir)"
 	@if "$(SHLIBS)"=="" \
-	 "$(PERL)" "$(SRCDIR)\util\copy.pl" ossl_static.pdb "$(libdir)"
+        "$(PERL)" "$(SRCDIR)\util\copy.pl" ossl_static.pdb "$(libdir)"
```
Comment (dot-asm) [inline Configurations/windows-makefile.tmpl]: This is unrelated modification, which break the formatting pattern.

**Comment by dot-asm** [comment]
> I [for one] can't find the suggested modification to 10-main.cf appropriate, [because] just dropping another file naming convention on everybody is hardly appropriate.

Apparently I failed to recognize that those were all new targets, so it's not like current naming convention is put aside. Yet it still think that new targets belong in separate .conf file.

**Comment by ObiWahn** [comment]
Is it possible to put the options for a static/dynamic build into the conf-files or has it to be passed as option?

@dot-asm the other PR seems to be about unix/shared libraries only. Can you guide me how to reuse it on windows?

**Comment by dot-asm** [comment]
I meant that attribute **name** could and preferably should be reused. So it's not like suggestion is to take some code from unix-Makefile.tmpl. But it would be appropriate to add few words to the corresponding note in README, about what does it mean on Windows.

File: Configurations/windows-makefile.tmpl [hunk 3]
```diff
@@ -3,14 +3,16 @@
...
  our $depext = $target{dep_extension} || ".d";
  our $exeext = $target{exe_extension} || ".exe";
- our $libext = $target{lib_extension} || ".lib";
- our $shlibext = $target{shared_extension} || ".dll";
- our $shlibextimport = $target{shared_import_extension} || ".lib";
- our $dsoext = $target{dso_extension} || ".dll";
+ our $libext = $lib_suffix . ($target{lib_extension} || ".lib");
+ our $shlibext = $lib_suffix . ($target{shared_extension} || ".dll");
+ our $shlibextimport = $lib_suffix . ($target{shared_import_extension} || ".lib");
+ our $dsoext = $lib_suffix . ($target{dso_extension} || ".dll");
```
Comment (dot-asm) [inline Configurations/windows-makefile.tmpl]: I haven't tested this, but if this line means that even engine dlls will get addendum in its name, then one has to recognize that when you ask to load engine X, the name is composed according to platform-specific predefined rules, so there is no place for name customization...

Comment (ObiWahn) [reply Configurations/windows-makefile.tmpl]: @dot-asm Could you elaborate a bit more on your concern? I am not familiar with your build chain and I have not messed with library names before. I this about the argument you need to pass to `dlopen()`? Is this [patch](https://www.npcglib.org/~stathis/downloads/openssl-1.1.0f.patch) going a better way? Please be a bit more explicit about what needs to be changed.

I am not sure what you want to say with the following statement
_So it's not like_ **the** / **a** _suggestion **(** is **)** to take some code from unix-Makefile.tmpl_ Depending on the missing word the meaning changes dramatically. Judging by the following sentence the first variant is more likely and you want the `$lib_suffix` to be more documented?! I think this can only be done if the technical part of this PR is set. Should it be put in the unix makefile template or not. How will the modifications work on Windows eventually?

**Comment by levitte** [comment]
I had forgotten about #4760.  Looking back at it, though, I *do* think that stealing the changes from `unix-Makefile.tmpl` would be quite appropriate.  Since `windows-makefile.tmpl` has a similar `shlib` function, it should be fairly easy, even...  Such a solution would be much more appropriate than to hack the extensions.

**Comment by levitte** [comment]
.. stealing *and adapting* ...

File: Configurations/windows-makefile.tmpl (hunk 3 above)
Comment (dot-asm) [reply Configurations/windows-makefile.tmpl]: Engine load request is meant to be OS-neutral. In sense that you'll use same call on all OS, and libcrypto till take care of OS specifics. If you ask load engine X, on Windows it will attempt to load X.DLL. If you produce XMD.DLL, then attempt to load X is bound to fail, which would be inappropriate.

> what you want to say

Right, I'm not in good terms with English, especially articles :-) What I suggest is to re-use `shlib_variant` attribute name (introduced earlier for similar kind of thing). I didn't have close look and rather assumed that taking snippets straight from unix-Makefile.tmpl wouldn't work. Richard seems to come to different conclusion, so this ought to mean that he looked closer. But taken code from unix-Makefile.tmpl or not, it would be only appropriate to add few words to Configurations/README about that `shlib_variant` is significant even on Windows and what does do there.

**Comment by levitte** [comment]
For `windows-makefile.tmpl`, the following diff *should* be sufficient:

``` diff
diff --git a/Configurations/windows-makefile.tmpl b/Configurations/windows-makefile.tmpl
index 11acdf00f6..f70c773a1a 100644
--- a/Configurations/windows-makefile.tmpl
+++ b/Configurations/windows-makefile.tmpl
@@ -8,6 +8,7 @@
  our $depext = $target{dep_extension} || ".d";
  our $exeext = $target{exe_extension} || ".exe";
  our $libext = $target{lib_extension} || ".lib";
+ our $shlibvariant = $target{shlib_variant} || "";
  our $shlibext = $target{shared_extension} || ".dll";
  our $shlibextimport = $target{shared_import_extension} || ".lib";
  our $dsoext = $target{dso_extension} || ".dll";
@@ -35,7 +36,7 @@
      my $lib = shift;
      return () if $disabled{shared} || $lib =~ /\.a$/;
      return () unless defined $unified_info{sharednames}->{$lib};
-     return $unified_info{sharednames}->{$lib} . $shlibext;
+     return $unified_info{sharednames}->{$lib} . $shlibvariant . $shlibext;
  }
 
  sub lib {
```

Then, all that's needed in appropriate config targets is something like this:

```
    shlib_variant => "MT",
```

Try it.

File: Configurations/windows-makefile.tmpl [hunk 4]
```diff
@@ -354,7 +356,7 @@ uninstall: uninstall_docs uninstall_sw
 
 libclean:
 	"$(PERL)" -e "map { m/(.*)\.dll$$/; unlink glob """{.,apps,test,fuzz}/$$1.*"""; } @ARGV" $(SHLIBS)
-	-del /Q /F $(LIBS) libcrypto.* libssl.* ossl_static.pdb
+	-del /Q /F $(LIBS) libcrypto*.* libssl*.* ossl_static*.pdb
```
Comment (levitte) [inline Configurations/windows-makefile.tmpl]: We *could* make better use of `$(SHLIBS)` and `$(SHLIBPDBS)`, no?

Comment (ObiWahn) [reply Configurations/windows-makefile.tmpl]: Should `shlib_variant` be renamed to `lib_variant` as at least on windows it meant to change the name of the static version as well? Is dynamic loading used within the OpenSSL project itself? If the engine dll's do not have different names then how can they co-exist in the same directory? By looking at the [FindOpenSSL.cmake](https://github.com/Kitware/CMake/blob/master/Modules/FindOpenSSL.cmake#L152) at least for debug and release the same paths are expected and lib/dll names differ in the suffix: "d" for the debug variant and "" for release. I do not like the CMake module at all. I would really prefer 4 different sub-folders with a fix naming scheme where each contains an unmodified OpenSSL install. But a change there will probably break far more projects than adding an additional OpenSSL build target. For me it is not clear what the ideal directory structure would look like and what the names of the different variants would be.
Maybe if the OpenSSL Project defines official locations and a structure on windows, it is enough for people that offer precompiled libraries and the CMake project to follow.  

I try to find a bit more time to learn more about linking and library resolving especially on windows. Thank you for pushing me in the right direction.

**Comment by levitte** [comment]
I just has a look at https://github.com/Kitware/CMake/blob/master/Modules/FindOpenSSL.cmake, and oh my what a pile of ideas, *none* of which correspond to the stuff our 1.1.0 builds produce (strangely, though, the pre-1.1.0 ones would be picked up all right)!  It's obviously been crafted around some binary redistribution of OpenSSL with its own ideas applied on top (the stathis stuff, I suspect).

There are more problems with that cmake recipe.  It's not just `_OPENSSL_MSVC_RT_MODE` that doesn't correspond to anything we produce, but `_OPENSSL_MSVC_ARCH_SUFFIX` is problematic as well, and is incompatible with the arch specific endings we normally use on 64bit (`-ia64` on Itanium and `-x64` on Amd64).

FindOpenSSL.cmake would not need much change to support our builds as well...  and quite frankly, I'm not entirely keen on adapting to some specific redistribution just because the Cmake kit did.  But that's my personal view, others (such as @dot-asm) might have a more positive view on this.

**Comment by dot-asm** [comment]
> we can as well generate static *.lib (naturally with distinct name) and then use it to link .dll (naturally with corresponding export .lib)

Just in case, it does work. Even passing static .lib alone, as opposite to passing long list of .obj files that is.

> [Difference between shared and non-shared builds would be absence of .dll and export library in former.]

Well, it would probably be appropriate to keep same name for "main" .lib file. I mean non-shared build outcome could be static libcrypto.lib, while shared build outcome would be export libcrypto.lib, .dll and additional uniquely named static .lib.
